pkg/controllers/clusterresourceplacement/resource_selector.go

```diff
@@ -131,38 +169,56 @@ func (r *Reconciler) gatherSelectedResource(placement string, selectors []fleetv
 
 func sortResources(resources []*unstructured.Unstructured) {
 	sort.Slice(resources, func(i, j int) bool {
+		// build the ordering map.
+		ordering := make(map[string]int, len(ApplyOrder))
+		for v, k := range ApplyOrder {
+			ordering[k] = v
+		}
+
 		obj1 := resources[i]
 		obj2 := resources[j]
-		gvk1 := obj1.GetObjectKind().GroupVersionKind().String()
-		gvk2 := obj2.GetObjectKind().GroupVersionKind().String()
-		// compare group/version;kind for the rest of type of resources
-		gvkComp := strings.Compare(gvk1, gvk2)
-		if gvkComp == 0 {
-			// same gvk, compare namespace/name, no duplication exists
-			return strings.Compare(fmt.Sprintf("%s/%s", obj1.GetNamespace(), obj1.GetName()),
-				fmt.Sprintf("%s/%s", obj2.GetNamespace(), obj2.GetName())) > 0
-		}
-		// sort by the cluster scoped priority resource types first
-		if gvk1 == utils.NamespaceMetaGVK.String() || gvk2 == utils.NamespaceMetaGVK.String() {
-			return gvk1 == utils.NamespaceMetaGVK.String()
-		}
-		if gvk1 == utils.CRDMetaGVK.String() || gvk2 == utils.CRDMetaGVK.String() {
-			return gvk1 == utils.CRDMetaGVK.String()
+		k1 := obj1.GetObjectKind().GroupVersionKind().Kind
+		k2 := obj2.GetObjectKind().GroupVersionKind().Kind
+
+		first, aok := ordering[k1]
+		second, bok := ordering[k2]
+		// if both kinds are unknown.
+		if !aok && !bok {
+			return lessByGVK(obj1, obj2, false)
 		}
-		// followed by namespaced priority resource types
-		if gvk1 == utils.ConfigMapGVK.String() || gvk2 == utils.ConfigMapGVK.String() {
-			return gvk1 == utils.ConfigMapGVK.String()
+		// unknown kind is last.
+		if !aok {
+			return false
 		}
-		if gvk1 == utils.SecretGVK.String() || gvk2 == utils.SecretGVK.String() {
-			return gvk1 == utils.SecretGVK.String()
+		if !bok {
+			return true
 		}
-		if gvk1 == utils.PersistentVolumeClaimGVK.String() || gvk2 == utils.PersistentVolumeClaimGVK.String() {
-			return gvk1 == utils.PersistentVolumeClaimGVK.String()
+		// same kind.
+		if first == second {
+			return lessByGVK(obj1, obj2, true)
 		}
-		return gvkComp < 0
+		// different known kinds, sort based on order index.
+		return first < second
 	})
 }
 
+func lessByGVK(obj1, obj2 *unstructured.Unstructured, ignoreKind bool) bool {
+	var gvk1, gvk2 string
+	if !ignoreKind {
```

Comment:
Hi Arvind! Just a nit here: could we switch the two branches just for a bit better readability?